### DISCUSSION TITLE
Merging Development Updates to Main

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#56] [General]
**Just FYI:** I'm going to start preparing to merge all of the development changes that we have been working on over the last several months back into the Main branch.  This will bring a lot of great new features and capabilities to the main release.  I've started to work on the auto-installer to update with a couple of new features.  I will also start working on the documentation and getting that updated to reflect the new features and install process.  I'm excited to get this updated and pushed out to the community.  If there are any links to 3D Printable Files, PCB design updates or companion smart phone apps that you feel should be incorporated, please let me know here.  (hint.. hint... @weberbox)

We can still work on driving the new development on the development branch in parallel as usual. 

Thanks!!  
Ben.

**Comment by weberbox** [comment]
Ok here are all the links I can think of right now. 


3D Printed part pictures
https://github.com/nebhead/PiFire/discussions/28#discussioncomment-1170748 

The below 3D printed items were designed for a PitBoss FB700 but could probably be used on other grills. The OnShape links are also provided so the designs can be modified to suite specific needs. 

Display Box Thingiverse
https://www.thingiverse.com/thing:5139144

Display Box OnShape
https://cad.onshape.com/documents/7c30dabaf483914d5d0ea10d/w/34b11bc17b55cc9ec2cbd3f8/e/681017a897d266fa8a7bcd75?renderMode=0&uiState=61945d0df20f441b9e619841

Control Box Thingiverse
https://www.thingiverse.com/thing:5139175 

Control Box OnShape
https://cad.onshape.com/documents/5e8e29627182ca9be4d6c313/w/8144537579abfcb512d691ce/e/f9ef421aba3cd2a53878154b?renderMode=0&uiState=619465dd0781005a1b58ebcc 

Auger Spacer OnShape (I needed this as my auger fan started to touch the control box) stl is with the control box on thingiverse
https://cad.onshape.com/documents/d1e81c5851a054343906a9ee/w/4cf9d91b67db4df4ce6475d6/e/35123253e2bf21c487d6c4cd?renderMode=0&uiState=619467e3b333ee4deac2a02a 

Distance Sensor Thingiverse
https://www.thingiverse.com/thing:5139180 

Distance Sensor OnShape
https://cad.onshape.com/documents/b95282d5ebd233bf22c48187/w/867e1e2e2f53551dd2575f1a/e/eaa8a945102367392b414161?renderMode=0&uiState=619467293958f85a7b286a18 

Distance Sensor Pics
https://github.com/nebhead/PiFire/discussions/17#discussioncomment-1493316 


EasyEda 4 Probe PCB Version
https://easyeda.com/zipster85/pifire-controller-hat 

Same basic design as the Nebhead PCB but with a header for buttons, a 4th input to the ADC and the mounting holes were enlarged a bit. This design was created from the Nebhead PCB. 

EasyEda Button PCB 
https://easyeda.com/zipster85/pifire-buttons 

The button PCB was designed for pull ups (HIGH for buttonslevel in common.py)


For mobile development PiFire Android v1.3.1 is designed for the current state of the PiFire development branch. (What I have been labeling v1.2.1 just so I can keep track of what I have installed, if you plan to use versions please feel free to change it to whatever you want. I can update the Android app side if you do want to change it) 

I have been using the app for a while and it has been working great so far. There still may be some bugs here and there as I cannot test every device but I think it is stable enough. There is an error reporting feature built in if the app crashes it is up to the user if they want to share a crash report or not. It can also be disabled in app settings if desired. 

Releases
https://github.com/weberbox/PiFire-Android/releases

There are currently two versions that can be downloaded

Github Firebase – This supports push notifications directly to the application from PiFire. It was redesigned so it no longer requires custom compiling the Android source and signing up for Firebase etc. This uses a hosted aws server to facilitate communication from a PiFire Server(grill)  to Google Firebase then to the Android Client. You only need to enable it from the app once you have connected the app to a PiFire server. The middle service is currently setup using a free tier of hosting and with the current user base of PiFire it should remain in the free tier. If PiFire ever grows large this may need to be reworked to avoid a monthly cost on my part. This version requires the device to have Google play services installed. 

Github NonFirebase – This version does not support mobile push notifications in the Android app. Pushbullet, Pushover, IFTTT, etc are still supported by the server but PiFire Android will not receive push notifications directly. This version does not require Google play services installed. 

There is also one more version that can be self compiled in case anyone wants to do that. It should be straight forward. Just download the source and import the project to Android Studio. There usually will be a few prompts to download an NDK and/or SDK. This option would require some additional work for push notifications as it does not include all the needed Google play files or server links. 

Regarding PiFire auto installer, much of that script was also used in the Ant build post install script. So it would probably be good to include your changes there as well at some point. 

If you ever intend on having PiFire releases the Ant build scripts can help there. I use it constantly to update PiFire as I am developing and testing and also when updating to new versions on the physical grill. 

Instructions for Ant are pretty simple, if you are on Linux and have Ant installed just cd into the PiFire project root folder and type ant in terminal. It should auto build PiFire and create two .deb packages one level up that can be installed on a raspberry pi, or other system (other defaults to prototype for the modules)

Then just move the deb to a system where you want to install it and run sudo apt install ./xxx.deb or sudo apt –reinstall install ./xxx.deb if the version has not changed but you want to install new files etc when developing. 

If you have no desire to release versions via deb’s no worries people can still build it with Ant themselves if they want and if you decide to keep the Ant scripts in the repo. 

I have started a new project at work that is keeping me pretty busy so development is probably going to slow down a bit on the Android app for a while.

**Reply by nebhead** [reply]
Wow!  Thanks for all the information.  This will definitely help me in my quest to update the documentation.  

Good luck with the work project!

**Reply by nebhead** [reply]
One other thought... I am going to implement a quick/dirty backup and restore function for the settings and pellet files so that it is easier to do a rebuild/reinstall for those who might be transitioning from the older mainline version.  Probably put it into the admin page.

**Reply by weberbox** [reply]
Oh nice I like that great idea.

**Comment by SteveOrangatang** [comment]
I seem to be getting an internal server error after the install. Has anyone had this issue before? I am using a Raspberry Pi 3, but I don't think thats the issue. I was able to run through the entire setup and can see the supervisor, but when I got to the IP, it throws an internal server error.

**Reply by nebhead** [reply]
> I seem to be getting an internal server error after the install. Has anyone had this issue before? I am using a Raspberry Pi 3, but I don't think thats the issue. I was able to run through the entire setup and can see the supervisor, but when I got to the IP, it throws an internal server error.

Can you check the Supervisor UI and see if the scripts are running?  

http://your.pi.ip.addr:9001

If the webapp script crashed, you can check the logs by ssh'ing into your pi and going to the ~/pifire/logs folder.  You should see an webapp err log and out log.  Check those logs to see if you have any error messages.

**Reply by SteveOrangatang** [reply]
The supervisor UI is definitely running. I'm not super familiar with navigating the terminal. When I SSH in and type ~/pifire/logs all I get for a response is "is a directory". Is there something else I'm supposed to do?

**Reply by SteveOrangatang** [reply]
I figured it out. Here is the log:

  GNU nano 5.4                     webapp.err.log                               
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType
[2022-01-11 18:33:29,155] ERROR in app: Exception on /dash [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2073, in wsg>
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1518, in ful>
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1516, in ful>
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1502, in dis>
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/home/pi/pifire/app.py", line 52, in dash
    control = ReadControl()
  File "/home/pi/pifire/common.py", line 390, in ReadControl
    control = json.loads(cmdsts.get('control:general'))
  File "/usr/lib/python3.9/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType

**Reply by nebhead** [reply]
Thanks!  Haven't seen this issue before.  I wonder if something went wrong during the installation.  Can try checking the Redis database at the command prompt?  

```
pi@pifire:~ $ redis-cli
127.0.0.1:6379> get control:general
```

And let me know what your output looks like.

**Reply by SteveOrangatang** [reply]
Sure! this is what shows up. Probably not what you wanted :)

pi@raspberrypi:~ $ redis-cli
127.0.0.1:6379> get control:general
(nil)

**Reply by nebhead** [reply]
Ha!  I mean, that is not what should be coming back, but it probably means that the control script crashed.  Can you send the output from the control.err.log or control.out.log?  It's also in the ~/pifire/logs folder.  

```
$ cat ~/pifire/logs/control.err.log 
```

**Reply by SteveOrangatang** [reply]
Uploaded as a text file due to size. Just to clarify, I don't have it wired up yet, not sure if that is the reason. I assume the dashboard still works without it hooked up? 

[control log.txt](https://github.com/nebhead/PiFire/files/7849831/control.log.txt)

**Reply by weberbox** [reply]
> 
> Uploaded as a text file due to size. Just to clarify, I don't have it wired up yet, not sure if that is the reason. I assume the dashboard still works without it hooked up? 
> 
> [control log.txt](https://github.com/nebhead/PiFire/files/7849831/control.log.txt)
> 

By not wired up do you mean that you are just using only the raspberry pi with nothing attached at all? If so you will need to set the modules to prototype. From the log it cannot find the ads.

You can cd into the pifire folder and run the modules.sh script and set the adc to prototype. Same thing for pellet level sensor and display if nothing is attached to the pi.   

cd ~/pifire
./modules.sh

If the script will not run you might need to give it execution permissions 

chmod +x modules.sh

**Reply by nebhead** [reply]
Yep, your control script is crashing because the ADC isn't responding.  When the control script crashes, it doesn't get a chance to put the control information into the Redis DB, and then the WebUI chokes when you try to access it.  This is kind of expected behavior, as we don't have graceful error handling at the moment.  

If you don't have this all wired up yet and still want to play around with the dashboard, you can change some of the modules to prototype modules.  You can do this configuration by running the following: 

```
bash ~/pifire/modules.sh
```

Then following the prompts, select PiFire for the Grill Platform (since that's just setting up the Raspberry Pi), select prototype for the ADC module since that isn't wired up yet, if you don't have the display wired up, select prototype for the display, and select prototype for range sensor (hopper/pellet level sensor) if you don't have that wired up either.  Then you should be able to reboot the Pi and check out the dashboard.  

Just remember that later, when you get it wired up, you'll want to switch back to the real devices.  So just repeat the command above and select the real devices.

**Comment by SteveOrangatang** [comment]
Excellent. Will give it a try. Thanks so much for the quick response!
